### PR TITLE
Add SSH-KDF support to Engine

### PR DESCRIPTION
**Pull request by ccalik**
+ Add SSH-KDF support to Engine. SSH-KDF implementation is disabled by default and can be enabled by adding -DSCOSSL_SSHKDF=1 argument to CMake. OpenSSL source code is required in the build process.

### CODE REVIEW

File: SymCryptEngine/src/scossl_sshkdf.c [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    EVP_KDF_CTX *ctx;
+
+    if(id != EVP_KDF_SSHKDF) {
+        return NULL;
+    }
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    
+    if (!ctx) {
+
```
Comment (samuel-lee-msft) [inline SymCryptEngine/src/scossl_sshkdf.c:310]: NP: remove empty line

**Review by samuel-lee-msft — CHANGES_REQUESTED** [review]
Generally looks good to me 👍 - just a few nitpicks

File: SymCryptEngine/dynamic/CMakeLists.txt [hunk 2]
```diff
@@ -30,6 +31,16 @@ add_library(scossl_dynamic SHARED
     ../src/scossl_helpers.c
 )
 
+check_symbol_exists(EVP_KDF_SSHKDF "openssl/kdf.h" SCOSSL_SSHKDF)
+
+if (SCOSSL_SSHKDF)
+    set(SCOSSL_SOURCES ${SCOSSL_SOURCES} ../src/scossl_sshkdf.c)
```
Comment (mlindgren) [inline SymCryptEngine/dynamic/CMakeLists.txt:37]: >set(SCOSSL_SOURCES ${SCOSSL_SOURCES} ../src/scossl_sshkdf.c) [](http://example.com/codeflow?start=4&length=60)

Nit: normally we do this via `list(APPEND SCOSSL_SOURCES ...)`

File: SymCryptEngine/static/CMakeLists.txt [hunk 3]
```diff
@@ -28,6 +29,17 @@ add_library(scossl_static STATIC
     ../src/scossl_helpers.c
 )
 
+check_symbol_exists(EVP_KDF_SSHKDF "openssl/kdf.h" SCOSSL_SSHKDF)
+
+if (SCOSSL_SSHKDF)
+    set(SCOSSL_SOURCES ${SCOSSL_SOURCES} ../src/scossl_sshkdf.c)
```
Comment (mlindgren) [inline SymCryptEngine/static/CMakeLists.txt:35]: >set(SCOSSL_SOURCES ${SCOSSL_SOURCES} ../src/scossl_sshkdf.c) [](http://example.com/codeflow?start=4&length=60)

`list(APPEND ...)`

File: README.md [hunk 4]
```diff
@@ -16,11 +16,13 @@ has not been tested for use in other environments and should not be considered p
 ## Algorithms that will be routed to a FIPS certifiable SymCrypt module with this version
 
 The following list is not necessarily exhaustive, and will be updated as more functionality is added to SCOSSL.
-Note that just because an algorithm is FIPS certifiable, does not mean it is recommended for use.
+Note that just because an algorithm is FIPS certifiable, does not mean it is recommended for use. SSH-KDF implementation
+will only be built if OpenSSL supports it. This will also require OpenSSL source code as part of the build process.
```
Comment (samuel-lee-msft) [inline README.md:20]: This sentence is no longer accurate?

**Review by samuel-lee-msft — APPROVED** [review]

File: README.md (hunk 4 above)
Comment (ccalik) [reply README.md]: Updated.

**Pull request merged**